File: app/adcsComputeTest/airBearingDemo/Harbenger_CONTROL/Simulation - Test/Observer.m
classdef Observer < handle
	properties
        % Observer States
        states_hat
        P

        % Model
        C_m
        D_m
        G
        Q
        R
        I
        Ts
		
	end
	methods
        function self = Observer(parameters)
        % Observer States
            self.states_hat = [0;0;0;0;0;0];
            self.P          = parameters.P_0;

             % Model
            self.C_m = [zeros(3,3),eye(3)];
            self.D_m = parameters.D_m;

        % Observer gains
            self.G = parameters.G;
            self.Q = parameters.Q1;
            self.R = parameters.R1;

            self.Ts = parameters.sample;
            self.I = parameters.I;
			
		end
        function x = update(self, inputs, measurements)
			

			% Propigate the ekf forward in time
            self.propigate_ekf(inputs, measurements);

			% Check if there is any new sensor data
            if length(measurements)>3
                self.star_update_ekf(inputs, measurements(4:end));
            end
			 
            x = [MRP2quat(self.states_hat(1:3));self.states_hat(4:6)];

        end
        function [x_hat_dot,P_dot] = ekf_dif_eq(self, x, P, inputs, measurements)
%             % Unpack
%             q1 = x(1);
% 			q2 = x(2);
% 			q3 = x(3);
% 			q4 = x(4);
% 
% 			wx = x(5);
% 			wy = x(6);
% 			wz = x(7);
% 
%             Ixx = self.I(1,1);
% 			Iyy = self.I(2,2);
% 			Izz = self.I(3,3);

%             A = [...
% 			0, -wx/2, -wy/2, -wz/2,                 -q2/2,                -q3/2,                 -q4/2;
%             wx/2,     0,  wz/2, -wy/2,                  q1/2,                -q4/2,                  q3/2;
%             wy/2, -wz/2,     0,  wx/2,                  q4/2,                 q1/2,                 -q2/2;
%             wz/2,  wy/2, -wx/2,     0,                 -q3/2,                 q2/2,                  q1/2;
%             0,     0,     0,     0,                     0, (wz*(Iyy - Izz))/Ixx,  (wy*(Iyy - Izz))/Ixx;
%             0,     0,     0,     0, -(wz*(Ixx - Izz))/Iyy,                    0, -(wx*(Ixx - Izz))/Iyy;
%             0,     0,     0,     0,  (wy*(Ixx - Iyy))/Izz, (wx*(Ixx - Iyy))/Izz,                     0;
%             ];
            [A,~,~,~]= modrod_state_space_fast([MRP2quat(x(1:3));x(4:6)],inputs,self.I);
      
            u = inputs;
             % Nonlinear differential equations to get a more accurate
            % propigation phase.
			q = MRP2quat(x(1:3));
			omega = x(4:6);
            sigma = x(1:3);

			q_dot = 0.5 .* quatmultiply([0;omega]',q');
			I = self.I;
%             inputs_e = zeros(3,1);
%             x_e = zeros(7,1);
%             x_e(1) = 1;
%             [A,B,C,D] = modrod_state_space(x_e, inputs_e, I);

			omega_dot = inv(I) * (u - cross(omega, I * omega));
            
            %f = [q_dot(1:4);omega_dot];
            %q_dot to mrp_dot and add omega_dot to end
            %f = [((1 + q(1)) * q_dot(2:4) - q_dot(1) * q(2:4))/(1 + q(1))^2; omega_dot];
            sigma_dot = 1/4*((1-norm(sigma)^2)*eye(3) + 2*cross_mat(sigma) + 2*(sigma*sigma'))*omega;
            f = [sigma_dot;omega_dot];

            y_m = measurements(1:3);

            % Kalman Gain
            K_K = P*self.C_m'*self.R^-1;

            % Kalman differential equations.
            y_hat = self.C_m*x;% Technically X_hat

            x_hat_dot = f + K_K*(y_m - y_hat);

            P_dot = A*P + P*A' - P*self.C_m'*self.R^-1*self.C_m*P + self.G*self.Q*self.G';

        end
        function [x_hat,P] = star_update_ekf(self, inputs, measurements)
           

            P = self.P;
            %R = self.R(7:end,7:end); % Only pull off Star Tracker 
            R = self.R;

            y_m = quat2MRP(measurements);

            C = [eye(3),zeros(3,3)];

            y_hat = self.states_hat(1:3);

            P = P - P*C'*(C*P*C'+R)^(-1)*C*P;
            K_L = P*C'*R^-1;

            e_y =  y_m - y_hat;

            x_hat = self.states_hat + K_L * e_y;

            self.states_hat = x_hat;
            self.P = P;

            

        end
        function propigate_ekf(self, inputs, measurements)
            %% Propogate EKF Forwared
			%[x_hat,P] = Integator.integrate([x_hat_dot,P_dot],self.ts);
            
            [x_hat_dot_1, P_dot_1] = self.ekf_dif_eq(self.states_hat,                         self.P,                     inputs, measurements);
            [x_hat_dot_2, P_dot_2] = self.ekf_dif_eq(self.states_hat + self.Ts/2*x_hat_dot_1, self.P + self.Ts/2*P_dot_1, inputs, measurements);
            [x_hat_dot_3, P_dot_3] = self.ekf_dif_eq(self.states_hat + self.Ts/2*x_hat_dot_2, self.P + self.Ts/2*P_dot_2, inputs, measurements);
            [x_hat_dot_4, P_dot_4] = self.ekf_dif_eq(self.states_hat + self.Ts  *x_hat_dot_3, self.P + self.Ts  *P_dot_3, inputs, measurements);
            self.states_hat = self.states_hat + self.Ts/6 * (x_hat_dot_1 + 2*x_hat_dot_2 + 2*x_hat_dot_3 + x_hat_dot_4);
            self.P     = self.P     + self.Ts/6 * (    P_dot_1 + 2*    P_dot_2 + 2*    P_dot_3 +     P_dot_4);
        end
	end
end

File: app/adcsComputeTest/airBearingDemo/Harbenger_CONTROL/RustControl/src/observer.rs
use nalgebra as na;
use crate::parameters::Parameters;
use crate::utils::quaternion;

pub struct Observer {
    // Observer states
    states_hat: na::Vector6<f64>,
    p: na::Matrix6<f64>,

    // Model matrices
    c_m: na::Matrix3x6<f64>,
    d_m: na::Matrix3<f64>,
    g: na::Matrix6<f64>,
    q: na::Matrix6<f64>,
    r: na::Matrix3<f64>,
    inertia: na::Matrix3<f64>,
    ts: f64,
}

impl Observer {
    pub fn new(parameters: &Parameters) -> Self {
        Observer {
            // Initialize observer states
            states_hat: na::Vector6::zeros(),
            p: parameters.p_0.clone(),

            // Initialize matrices
            c_m: na::Matrix3x6::new(
                0.0, 0.0, 0.0, 1.0, 0.0, 0.0,
                0.0, 0.0, 0.0, 0.0, 1.0, 0.0,
                0.0, 0.0, 0.0, 0.0, 0.0, 1.0
            ),
            d_m: parameters.d_m.clone(),
            g: parameters.g.clone(),
            q: parameters.q1.clone(),
            r: parameters.r1.clone(),
            inertia: parameters.inertia.clone(),
            ts: parameters.sample,
        }
    }

    pub fn update(&mut self, inputs: &na::Vector3<f64>, measurements: &na::DVector<f64>) -> na::Vector7<f64> {
        // Propagate the EKF forward in time
        self.propagate_ekf(inputs, measurements);

        // Check if there is star tracker data
        if measurements.len() > 3 {
            // Extract star tracker measurements
            let star_measurements = na::Vector4::new(
                measurements[3], measurements[4], measurements[5], measurements[6]
            );
            
            self.star_update_ekf(inputs, &star_measurements);
        }

        // Convert current state estimate to quaternion form
        let mrp = na::Vector3::new(self.states_hat[0], self.states_hat[1], self.states_hat[2]);
        let quat = quaternion::mrp2quat(&mrp);
        
        // Combine quaternion and angular velocity into full state vector
        let mut x = na::Vector7::zeros();
        x[0] = quat[0];
        x[1] = quat[1];
        x[2] = quat[2];
        x[3] = quat[3];
        x[4] = self.states_hat[3];
        x[5] = self.states_hat[4];
        x[6] = self.states_hat[5];
        
        x
    }

    fn ekf_dif_eq(
        &self,
        x: &na::Vector6<f64>,
        p: &na::Matrix6<f64>,
        inputs: &na::Vector3<f64>,
        measurements: &na::DVector<f64>,
    ) -> (na::Vector6<f64>, na::Matrix6<f64>) {
        // Convert MRP to quaternion for linearization
        let mrp = na::Vector3::new(x[0], x[1], x[2]);
        let quat = quaternion::mrp2quat(&mrp);
        
        // Create full state vector for linearization
        let mut x_full = na::Vector7::zeros();
        x_full[0] = quat[0];
        x_full[1] = quat[1];
        x_full[2] = quat[2];
        x_full[3] = quat[3];
        x_full[4] = x[3];
        x_full[5] = x[4];
        x_full[6] = x[5];
        
        // Get linearized dynamics matrix
        let (a, _, _, _) = quaternion::modrod_state_space_fast(&x_full, inputs, &self.inertia);
        
        // Nonlinear differential equations for more accurate propagation
        let omega = na::Vector3::new(x[3], x[4], x[5]);
        let sigma = na::Vector3::new(x[0], x[1], x[2]);
        
        // Calculate MRP derivative
        let sigma_sq_norm = sigma.norm_squared();
        let identity = na::Matrix3::<f64>::identity();
        let cross_sigma = quaternion::cross_mat(&sigma);
        let sigma_sigma_t = sigma * sigma.transpose();
        
        let sigma_dot = 0.25 * ((1.0 - sigma_sq_norm) * identity + 
                               2.0 * cross_sigma + 
                               2.0 * sigma_sigma_t) * omega;
                               // Calculate angular acceleration
        let omega_dot = self.inertia.try_inverse().unwrap() * 
        (inputs - omega.cross(&(self.inertia * omega)));
    
    // Combine into state derivative
    let f = na::Vector6::new(
        sigma_dot[0], sigma_dot[1], sigma_dot[2],
        omega_dot[0], omega_dot[1], omega_dot[2]
    );
    
    // Extract gyroscope measurements
    let y_m = na::Vector3::new(measurements[0], measurements[1], measurements[2]);
    
    // Kalman gain
    let k_k = p * self.c_m.transpose() * self.r.try_inverse().unwrap();
    
    // Predicted output
    let y_hat = self.c_m * x;
    
    // State derivative
    let x_hat_dot = f + k_k * (y_m - y_hat);
    
    // Covariance derivative
    let p_dot = a * p + p * a.transpose() - 
                p * self.c_m.transpose() * self.r.try_inverse().unwrap() * self.c_m * p + 
                self.g * self.q * self.g.transpose();
    
    (x_hat_dot, p_dot)
}

fn star_update_ekf(&mut self, _inputs: &na::Vector3<f64>, measurements: &na::Vector4<f64>) {
    // Convert quaternion to MRP
    let mrp_measured = quaternion::quat2mrp(measurements);
    
    // Define observation matrix for star tracker (measures attitude directly)
    let c = na::Matrix3x6::new(
        1.0, 0.0, 0.0, 0.0, 0.0, 0.0,
        0.0, 1.0, 0.0, 0.0, 0.0, 0.0,
        0.0, 0.0, 1.0, 0.0, 0.0, 0.0
    );
    
    // Extract current attitude estimate
    let mrp_estimated = na::Vector3::new(
        self.states_hat[0], self.states_hat[1], self.states_hat[2]
    );
    
    // Update covariance
    let innovation_cov = c * self.p * c.transpose() + self.r;
    self.p = self.p - self.p * c.transpose() * innovation_cov.try_inverse().unwrap() * c * self.p;
    
    // Kalman gain
    let k_l = self.p * c.transpose() * self.r.try_inverse().unwrap();
    
    // Innovation (measurement error)
    let e_y = mrp_measured - mrp_estimated;
    
    // Update state estimate
    self.states_hat = self.states_hat + k_l * e_y;
}

fn propagate_ekf(&mut self, inputs: &na::Vector3<f64>, measurements: &na::DVector<f64>) {
    // RK4 integration for the EKF
    // Step 1
    let (x_hat_dot_1, p_dot_1) = self.ekf_dif_eq(
        &self.states_hat, &self.p, inputs, measurements
    );
    
    // Step 2
    let x_hat_2 = self.states_hat + x_hat_dot_1 * (self.ts / 2.0);
    let p_2 = self.p + p_dot_1 * (self.ts / 2.0);
    let (x_hat_dot_2, p_dot_2) = self.ekf_dif_eq(
        &x_hat_2, &p_2, inputs, measurements
    );
    
    // Step 3
    let x_hat_3 = self.states_hat + x_hat_dot_2 * (self.ts / 2.0);
    let p_3 = self.p + p_dot_2 * (self.ts / 2.0);
    let (x_hat_dot_3, p_dot_3) = self.ekf_dif_eq(
        &x_hat_3, &p_3, inputs, measurements
    );
    
    // Step 4
    let x_hat_4 = self.states_hat + x_hat_dot_3 * self.ts;
    let p_4 = self.p + p_dot_3 * self.ts;
    let (x_hat_dot_4, p_dot_4) = self.ekf_dif_eq(
        &x_hat_4, &p_4, inputs, measurements
    );
    
    // Combine steps
    self.states_hat = self.states_hat + 
        (x_hat_dot_1 + x_hat_dot_2 * 2.0 + x_hat_dot_3 * 2.0 + x_hat_dot_4) * (self.ts / 6.0);
    
    self.p = self.p + 
        (p_dot_1 + p_dot_2 * 2.0 + p_dot_3 * 2.0 + p_dot_4) * (self.ts / 6.0);
}
}